x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,428 @@
+            }
+        }
+
+        for (DelayedCategorizationBucket bucket : reduced.values()) {
+            // Parse log line takes document count into account and merging on smallest groups
+            categorizationTokenTree.parseLogLine(bucket.key.keyAsTokens(), bucket.docCount);
+        }
+        // Collapse tiny groups together, this may result in new bucket keys for already known buckets
+        categorizationTokenTree.mergeSmallestChildren();
+        Map<BucketKey, DelayedCategorizationBucket> mergedBuckets = new HashMap<>(aggregations.size(), 1.0f);
+        for (DelayedCategorizationBucket delayedBucket : reduced.values()) {
+            LogGroup group = categorizationTokenTree.parseLogLineConst(delayedBucket.key.keyAsTokens());
+            if (group == null) {
+                throw new AggregationExecutionException(
+                    "Unexpected null categorization group for bucket [" + delayedBucket.key.asString() + "]"
+                );
+            }
+            BucketKey key = new BucketKey(group.getLogEvent());
+            if (reduceContext.isFinalReduce()) {
+                key.collapseWildCards();
+            }
+            mergedBuckets.computeIfAbsent(
+                key,
+                k -> new DelayedCategorizationBucket(k, new ArrayList<>(delayedBucket.toReduce.size()), 0L)
+            ).add(delayedBucket);
+        }
+
+        final int size = reduceContext.isFinalReduce() == false ? buckets.size() : Math.min(requiredSize, buckets.size());
+        final PriorityQueue<Bucket> pq = new BucketCountPriorityQueue(size);
+        for (Map.Entry<BucketKey, DelayedCategorizationBucket> keyAndBuckets : mergedBuckets.entrySet()) {
+            final BucketKey key = keyAndBuckets.getKey();
+            DelayedCategorizationBucket bucket = keyAndBuckets.getValue();
+            Bucket newBucket = bucket.reduce(key, reduceContext);
+            if ((newBucket.docCount >= minDocCount) || reduceContext.isFinalReduce() == false) {
+                Bucket removed = pq.insertWithOverflow(newBucket);
+                if (removed == null) {
+                    reduceContext.consumeBucketsAndMaybeBreak(1);
+                } else {
+                    reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(removed));
+                }
+            } else {
+                reduceContext.consumeBucketsAndMaybeBreak(-countInnerBucket(newBucket));
+            }
+        }
+        Bucket[] bucketList = new Bucket[pq.size()];
+        for (int i = pq.size() - 1; i >= 0; i--) {
+            bucketList[i] = pq.pop();
+        }
+        return new InternalCategorizationAggregation(
+            name,
+            requiredSize,
+            minDocCount,
+            maxChildren,
+            maxDepth,
+            similarityThreshold,
+            metadata,
+            Arrays.asList(bucketList)
+        );
+    }
+
+    @Override
+    public Object getProperty(List<String> path) {
+        // TODO anything special?
```

Comment:
It's for pipeline aggs. So if you can think of anything special then go for it.

Reply:
Yeah, I might just blow this away, there isn't anything special here :D.